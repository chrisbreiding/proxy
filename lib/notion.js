const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
