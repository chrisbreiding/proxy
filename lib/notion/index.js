const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
