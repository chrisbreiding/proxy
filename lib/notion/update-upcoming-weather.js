const dayjs = require('dayjs')
const debug = require('debug')('proxy:scripts')

const { getData: getWeatherData } = require('../weather')
const {
  dateRegex,
  getDateFromText,
  getPlainText,
  getQuests,
  updateBlock,
  getWeatherIcon,
  makeTextPart,
} = require('./util')
const { compact } = require('../util/collections')
const { getEnv } = require('../util/env')

const notionToken = getEnv('NOTION_TOKEN')
const questsId = getEnv('NOTION_QUESTS_ID')
const weatherLocation = getEnv('WEATHER_LOCATION')
const dryRun = getEnv('DRY_RUN')

if (dryRun) {
  // eslint-disable-next-line no-console
  console.log('--- DRY RUN ---')
}

debug('ENV:', {
  notionToken,
  questsId,
  weatherLocation,
  dryRun,
})

const getDates = (questBlocks) => {
  return questBlocks.reduce((memo, block) => {
    if (block.type !== 'paragraph') return memo

    const text = getPlainText(block)
    const [dateText] = text.match(dateRegex) || []

    if (!dateText) return memo

    const date = getDateFromText(dateText).format('YYYY-MM-DD')

    memo.push({
      text,
      date,
      dateText,
      id: block.id,
    })

    return memo
  }, [])
}

const getWeather = async () => {
  const weather = await getWeatherData({ location: weatherLocation })

  return weather.daily.data.reduce((memo, dayWeather) => {
    const date = dayjs.unix(dayWeather.time).format('YYYY-MM-DD')

    debug('weather: %o', { timestamp: dayWeather.time, date })

    memo[date] = dayWeather

    return memo
  }, {})
}

const makePrecipPart = (condition, info) => {
  return condition ? makeTextPart(`(${info}) `, 'gray') : undefined
}

const updateBlockWeather = async (block, weather) => {
  const content = {
    paragraph: {
      rich_text: compact([
        makeTextPart(`${block.dateText}    ${getWeatherIcon(weather.icon)} `),
        makePrecipPart(weather.icon === 'rain' && weather.precipProbability >= 0.01, `${Math.round(weather.precipProbability * 100)}%`),
        makePrecipPart(weather.icon === 'snow' && weather.precipAccumulation >= 0.1, `${(weather.precipAccumulation || 0).toFixed(2)}in`),
        makeTextPart(`${Math.round(weather.temperatureLow)}°`, 'blue'),
        makeTextPart(' / ', 'gray'),
        makeTextPart(`${Math.round(weather.temperatureHigh)}°`, 'orange'),
      ]),
    },
  }

  const newText = content.paragraph.rich_text
  .map(({ text }) => text.content)
  .join('')
  .trim()

  if (newText === block.text) {
    debug(`No weather change for ${block.dateText}`)

    return
  }

  // eslint-disable-next-line no-console
  const log = dryRun && !debug.enabled ? console.log : debug

  log(`Update "${block.text}" to "${newText}"`)

  if (dryRun) return

  await updateBlock({ notionToken, block: content, blockId: block.id })
}

const updateBlocks = async (blocks, weather) => {
  for (let block of blocks) {
    if (weather[block.date]) {
      await updateBlockWeather(block, weather[block.date])
    } else {
      debug('No weather found for %s', block.dateText)
    }
  }
}

const updateWeather = async () => {
  try {
    const questBlocks = await getQuests({ notionToken, pageId: questsId })
    const dateBlocks = getDates(questBlocks)

    debug('dateBlocks:', dateBlocks)

    const weather = await getWeather({ location: weatherLocation })

    await updateBlocks(dateBlocks, weather)

    // eslint-disable-next-line no-console
    console.log('Quests successfully updated with weather')

    process.exit(0)
  } catch (error) {
    // eslint-disable-next-line no-console
    console.log('Updating quest weather failed:')
    // eslint-disable-next-line no-console
    console.log(error.stack)

    process.exit(1)
  }
}

updateWeather()
