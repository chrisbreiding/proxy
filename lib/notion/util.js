const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
