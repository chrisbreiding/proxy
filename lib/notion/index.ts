import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
