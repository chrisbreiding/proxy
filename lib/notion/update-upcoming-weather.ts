import dayjs from 'dayjs'

import {
  dateRegex,
  getBlockPlainText,
  getDateFromText,
  makeTextPart,
  NotionBlock,
  updateBlock,
} from './util'
import { getAllQuests } from './quests'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { makeConditionParts, makeTemperatureParts } from './weather'
import { DayWeather, getWeatherData } from '../weather'

interface DateObject {
  date: string
  dateText: string
  id: string
  text: string
}

function getDates (questBlocks: NotionBlock[]) {
  return questBlocks.reduce((memo, block) => {
    if (block.type !== 'paragraph') return memo

    const text = getBlockPlainText(block)

    if (!text) return memo

    const [dateText] = text.match(dateRegex) || []

    if (!dateText) return memo

    const date = getDateFromText(dateText).format('YYYY-MM-DD')

    memo.push({
      text,
      date,
      dateText,
      id: block.id,
    })

    return memo
  }, [] as DateObject[])
}

export type WeatherByDate = { [key: string]: DayWeather }

export function getWeatherByDate (weather: DayWeather[]) {
  return weather.reduce((memo, dayWeather) => {
    const date = dayjs.unix(dayWeather.time).format('YYYY-MM-DD')

    debugVerbose('weather: %o', { timestamp: dayWeather.time, date })

    memo[date] = dayWeather

    return memo
  }, {} as WeatherByDate)
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface UpdateBlockWeatherOptions {
  dateObject: DateObject
  notionToken: string
  weather: DayWeather
}

async function updateBlockWeather ({ dateObject, notionToken, weather }: UpdateBlockWeatherOptions) {
  const block = {
    type: 'paragraph' as const,
    content: {
      rich_text: [
        makeTextPart(`${dateObject.dateText}    `), // Mon, 11/1
        ...makeConditionParts(weather), // ☀️ | ☔️ (82%)
        ...makeTemperatureParts(weather), // 33° / 62°
      ],
      color: 'default' as const,
    },
  }

  const newText = block.content.rich_text
  .map(({ text }) => text.content)
  .join('')
  .trim()

  if (newText === dateObject.text) {
    debugVerbose(`No weather change for ${dateObject.dateText}`)

    return
  }

  debugVerbose(`Update "${dateObject.text}" to "${newText}"`)

  await updateBlock({ notionToken, block, blockId: dateObject.id })
}

interface UpdateBlocksOptions {
  dateObjects: DateObject[]
  notionToken: string
  weather: WeatherByDate
}

async function updateBlocks ({ dateObjects, notionToken, weather }: UpdateBlocksOptions) {
  for (const dateObject of dateObjects) {
    if (weather[dateObject.date]) {
      await updateBlockWeather({ dateObject, notionToken, weather: weather[dateObject.date] })
    /* c8 ignore next 3 */
    } else {
      debugVerbose('No weather found for %s', dateObject.dateText)
    }
  }
}

interface UpdateWeatherOptions {
  notionToken: string
  questsId: string
  location: string
}

export async function updateWeather ({ notionToken, questsId, location }: UpdateWeatherOptions) {
  const questBlocks = await getAllQuests({ notionToken, pageId: questsId })
  const dateObjects = getDates(questBlocks)

  debugVerbose('dateObjects:', dateObjects)

  const weatherData = await getWeatherData(location)
  const weather = await getWeatherByDate(weatherData.daily.data)

  await updateBlocks({ dateObjects, notionToken, weather })
}

export default async function main () {
  const notionToken = getEnv('NOTION_TOKEN')!
  const questsId = getEnv('NOTION_QUESTS_ID')!
  const location = getEnv('WEATHER_LOCATION')!

  debugVerbose('ENV:', {
    notionToken,
    questsId,
    location,
  })

  try {
    debug('Updating quests with weather...')

    await updateWeather({ notionToken, questsId, location })

    debug('Successfully updated quests with weather')
  } catch (error: any) {
    debug('Updating quest weather failed:')
    debug(error?.stack || error)
  }
}
