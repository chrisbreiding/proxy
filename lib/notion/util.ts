import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
