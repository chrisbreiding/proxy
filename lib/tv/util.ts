import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
