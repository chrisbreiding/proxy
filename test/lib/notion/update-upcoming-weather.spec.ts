import dayjs from 'dayjs'
import nock from 'nock'
import { afterEach, beforeEach, describe, it, vi } from 'vitest'

process.env.DARK_SKY_API_KEY = 'dark-sky-key'

import { updateWeather } from '../../../lib/notion/update-upcoming-weather'
import { snapshotBody } from '../../util'
import {
  block,
  nockGetBlockChildren,
  nockNotion,
  notionFixture as fixture,
  notionFixtureContents as fixtureContents,
} from './util'

describe('lib/notion/update-upcoming-weather', () => {
  beforeEach(() => {
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  it('updates upcoming quest dates with temperature and conditions', async () => {
    vi.setSystemTime(new Date(2022, 11, 1)) // just needs to be before 12/27

    const weather = fixtureContents('weather/weather')

    weather.daily.data[0].time = dayjs('2022-12-27').unix()
    weather.daily.data[1].time = dayjs('2022-12-28').unix()
    weather.daily.data[2].time = dayjs('2022-12-29').unix()
    weather.daily.data[3].time = dayjs('2022-12-30').unix()
    weather.daily.data[4].time = dayjs('2022-12-31').unix()
    weather.daily.data[5].time = dayjs('2023-01-01').unix()
    weather.daily.data[6].time = dayjs('2023-01-02').unix()

    nock('https://api.darksky.net')
    .get('/forecast/dark-sky-key/lat,lng?exclude=minutely,flags&extend=hourly')
    .reply(200, weather)

    const questBlocks = [
      block.p({ text: 'Some text' }),
      block.p({ text: '' }),
      block.p({ text: 'Mon, 12/27    ☀️ 33° / 62°' }),
      block.bullet({ text: 'A task' }),
      block.bullet({ text: 'A task' }),
      block.p({ text: 'Tue, 12/28' }),
      block.bullet({ text: 'A task' }),
      block.toggle({ text: 'Upcoming', id: 'upcoming-id' }),
    ]

    const upcomingBlocks = [
      block.p({ text: 'Wed, 12/29' }),
      block.bullet({ text: 'A task' }),
      block.p({ text: 'Thu, 12/30' }),
      block.bullet({ text: 'A task' }),
      block.p({ text: 'Fri, 12/31' }),
      block.bullet({ text: 'A task' }),
      block.p({ text: 'Sat, 1/1' }),
      block.bullet({ text: 'A task' }),
      block.p({ text: 'Sun, 1/2' }),
      block.bullet({ text: 'A task' }),
    ]

    nockGetBlockChildren('quests-id', { reply: { results: questBlocks } })
    nockGetBlockChildren('upcoming-id', { reply: { results: upcomingBlocks } })

    const snapshotUpdates = [6, 8, 10, 12, 14, 16].map((num) => {
      return snapshotBody(nockNotion({
        method: 'patch',
        path: `/v1/blocks/block-${num}`,
      }), `block-${num}-update`)
    })

    await updateWeather({
      notionToken: 'notion-token',
      questsId: 'quests-id',
      location: 'lat,lng',
    })

    await Promise.all(snapshotUpdates)
  })
})
