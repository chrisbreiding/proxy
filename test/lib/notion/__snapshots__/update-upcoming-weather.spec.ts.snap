// Vitest Snapshot v1

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-6-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Tue, 12/28    ",
        },
      },
      {
        "text": {
          "content": "⛅️ ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "45°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "66°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-8-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Wed, 12/29    ",
        },
      },
      {
        "text": {
          "content": "⛅️ ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "43°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "66°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-10-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Thu, 12/30    ",
        },
      },
      {
        "text": {
          "content": "☔️ ",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": "(85%) ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "33°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "54°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-12-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Fri, 12/31    ",
        },
      },
      {
        "text": {
          "content": "❄️ ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "25°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "51°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-14-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Sat, 1/1    ",
        },
      },
      {
        "text": {
          "content": "⛅️ ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "26°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "51°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;

exports[`lib/notion/update-upcoming-weather > updates upcoming quest dates with temperature and conditions > block-16-update 1`] = `
{
  "object": "block",
  "paragraph": {
    "color": "default",
    "rich_text": [
      {
        "text": {
          "content": "Sun, 1/2    ",
        },
      },
      {
        "text": {
          "content": "⛅️ ",
        },
      },
      {
        "annotations": {
          "color": "blue",
        },
        "text": {
          "content": "29°",
        },
      },
      {
        "annotations": {
          "color": "gray",
        },
        "text": {
          "content": " / ",
        },
      },
      {
        "annotations": {
          "color": "orange",
        },
        "text": {
          "content": "63°",
        },
      },
    ],
  },
  "type": "paragraph",
}
`;
